src/Silverback.Integration/Messaging/Configuration/EndpointsConfigurationBuilderAddOutboundExtensions.cs

```diff
@@ -183,22 +220,44 @@ public static class EndpointsConfigurationBuilderAddOutboundExtensions
         ///     Specifies whether the producers must be immediately instantiated and connected. When <c>false</c> the
         ///     <see cref="IProducer" /> will be created only when the first message is about to be produced.
         /// </param>
+        /// <param name="idempotentRegistration">
+        ///     Only adds the endpoint when it is not already registered.
+        ///     <remarks>
+        ///         Validates only against static routes based on the message type (see <see cref="StaticOutboundRouter"/>),
+        ///         not routes that are calculated at runtime based on the message.
+        ///     </remarks>
+        /// </param>
         /// <returns>
         ///     The <see cref="IEndpointsConfigurationBuilder" /> so that additional calls can be chained.
         /// </returns>
         public static IEndpointsConfigurationBuilder AddOutbound(
             this IEndpointsConfigurationBuilder endpointsConfigurationBuilder,
             Type messageType,
             IEnumerable<IProducerEndpoint> endpoints,
-            bool preloadProducers = true)
+            bool preloadProducers = true,
+            bool idempotentRegistration = false)
         {
             Check.NotNull(endpointsConfigurationBuilder, nameof(endpointsConfigurationBuilder));
             Check.NotNull(messageType, nameof(messageType));
             Check.NotNull(endpoints, nameof(endpoints));
 
+            IEnumerable<IProducerEndpoint> endpointsToRegister = endpoints;
+
+            if (idempotentRegistration)
+            {
+                var registeredStaticEndpoints = endpointsConfigurationBuilder
+                    .GetOutboundRoutingConfiguration()
+                    .GetRoutesForMessage(messageType)
+                    .Select(route => route.GetOutboundRouter(endpointsConfigurationBuilder.ServiceProvider))
+                    .OfType<StaticOutboundRouter>()
+                    .SelectMany(router => router.Endpoints);
+
+                endpointsToRegister = endpoints.Except(registeredStaticEndpoints);
```

Comment:
I wonder if this isn't a bit fragile.

I mean, since we do compare the entire endpoint configuration, it's enough that you change a setting in one of the features to lead to a duplicate.

You should only check for matching message type / endpoint name pairs.
If the endpoint is about to be discarded then you check whether the entire configuration matches with the already registered one and I'd expect at least a warning to be logged in that case. I think you could simply throw an exception and let it break, it's a configuration error.